src/machines/systemIO/systemIOMachine.ts

```diff
@@ -0,0 +1,448 @@
+        input: ({ context, event, self }) => {
+          assertEvent(event, SystemIOMachineEvents.createKCLFile)
+          return {
+            context,
+            requestedProjectName: event.data.requestedProjectName,
+            requestedFileName: event.data.requestedFileName,
+            requestedCode: event.data.requestedCode,
+            rootContext: self.system.get('root').getSnapshot().context,
+          }
+        },
+        onDone: {
+          target: SystemIOMachineStates.idle,
+        },
+        onError: {
+          target: SystemIOMachineStates.idle,
+          actions: [SystemIOMachineActions.toastError],
+        },
+      },
+    },
+    [SystemIOMachineStates.importFileFromURL]: {
+      invoke: {
+        id: SystemIOMachineActors.importFileFromURL,
+        src: SystemIOMachineActors.createKCLFile,
+        input: ({ context, event, self }) => {
+          assertEvent(event, SystemIOMachineEvents.importFileFromURL)
+          return {
+            context,
+            requestedProjectName: event.data.requestedProjectName,
+            requestedFileName: event.data.requestedFileName,
+            requestedCode: event.data.requestedCode,
+            rootContext: self.system.get('root').getSnapshot().context,
+          }
+        },
+        onDone: {
+          target: SystemIOMachineStates.readingFolders,
+          // Clear on web? not desktop
+          actions: [
+            assign({
+              requestedFileName: ({ context, event }) => {
+                assertEvent(event, SystemIOMachineEvents.done_importFileFromURL)
+                // Not the entire path
+                return {
+                  project: event.output.projectName,
+                  file: event.output.fileName + '.kcl',
+                }
+              },
+            }),
+            assign({ clearURLParams: { value: true } }),
+          ],
+        },
+        onError: {
+          target: SystemIOMachineStates.idle,
+          actions: [SystemIOMachineActions.toastError],
+        },
+      },
+    },
+    [SystemIOMachineStates.checkingReadWrite]: {
+      invoke: {
+        id: SystemIOMachineActors.checkReadWrite,
+        src: SystemIOMachineActors.checkReadWrite,
+        input: ({ context, event }) => {
+          assertEvent(event, SystemIOMachineEvents.setProjectDirectoryPath)
+          return {
+            context,
+            requestedProjectDirectoryPath:
+              event.data.requestedProjectDirectoryPath,
+          }
+        },
+        onDone: {
+          target: SystemIOMachineStates.readingFolders,
+        },
+        onError: {
+          target: SystemIOMachineStates.readingFolders,
+          actions: [SystemIOMachineActions.toastError],
+        },
+      },
+    },
+  },
+})
+
+// Watcher handler
+// look at projectDirectory useEffect then send this event if it changes or if we need to do this?
+// The handler needs to live somewhere... aka the provider?
```

Comment:
Are these notes from while writing this maybe?

Reply:
notes, yeah removing.